pimd/pim_mlag.c

```diff
@@ -0,0 +1,346 @@
+		if (rc)
+			return rc;
+		pim_mlag_process_vxlan_update(&msg);
+	} break;
+	case MLAG_MROUTE_ADD: {
+		struct mlag_mroute_add msg;
+
+		rc = zebra_mlag_lib_decode_mroute_add(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_mroute_add(msg);
+	} break;
+	case MLAG_MROUTE_DEL: {
+		struct mlag_mroute_del msg;
+
+		rc = zebra_mlag_lib_decode_mroute_del(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_mroute_del(msg);
+	} break;
+	case MLAG_MROUTE_ADD_BULK: {
+		struct mlag_mroute_add msg;
+		int i = 0;
+
+		for (i = 0; i < mlag_msg.msg_cnt; i++) {
+
+			rc = zebra_mlag_lib_decode_mroute_add(s, &msg);
+			if (rc)
+				return (rc);
+			pim_mlag_process_mroute_add(msg);
+		}
+	} break;
+	case MLAG_MROUTE_DEL_BULK: {
+		struct mlag_mroute_del msg;
+		int i = 0;
+
+		for (i = 0; i < mlag_msg.msg_cnt; i++) {
+
+			rc = zebra_mlag_lib_decode_mroute_del(s, &msg);
+			if (rc)
+				return (rc);
+			pim_mlag_process_mroute_del(msg);
+		}
+	} break;
+	default:
+		break;
+	}
+	return 0;
+}
+
+/****************End of PIM Mesasge processing handler********************/
+
+int pim_zebra_mlag_process_up(void)
+{
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Received Process-Up from Mlag", __func__);
+
+	return 0;
+}
+
+int pim_zebra_mlag_process_down(void)
+{
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Received Process-Down from Mlag", __func__);
+
+	return 0;
+}
+
+static int pim_mlag_register_handler(struct thread *thread)
+{
+	uint32_t bit_mask = 0;
+
+	if (!zclient)
+		return -1;
+
+	SET_FLAG(bit_mask, (1 << MLAG_STATUS_UPDATE));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_ADD));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_DEL));
+	SET_FLAG(bit_mask, (1 << MLAG_DUMP));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_ADD_BULK));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_DEL_BULK));
+	SET_FLAG(bit_mask, (1 << MLAG_PIM_CFG_DUMP));
+	SET_FLAG(bit_mask, (1 << MLAG_VXLAN_UPDATE));
+	SET_FLAG(bit_mask, (1 << MLAG_PEER_FRR_STATUS));
+
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Posting Client Register to MLAG mask:0x%x",
+			   __func__, bit_mask);
+
+	zclient_send_mlag_register(zclient, bit_mask);
+	return 0;
+}
+
+void pim_mlag_register(void)
+{
+	if (router->mlag_process_register)
+		return;
+
+	router->mlag_process_register = true;
+
+	thread_add_event(router->master, pim_mlag_register_handler, NULL, 0,
+			 NULL);
+}
+
+static int pim_mlag_deregister_handler(struct thread *thread)
+{
+	if (!zclient)
+		return -1;
+
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Posting Client De-Register to MLAG from PIM",
+			   __func__);
+	router->connected_to_mlag = false;
+	zclient_send_mlag_deregister(zclient);
+	return 0;
+}
+
+void pim_mlag_deregister(void)
+{
+	/* if somebody still interested in the MLAG channel skip de-reg */
+	if (router->pim_mlag_intf_cnt)
+		return;
+
+	/* not registered; nothing do */
+	if (!router->mlag_process_register)
+		return;
+
+	router->mlag_process_register = false;
+
+	thread_add_event(router->master, pim_mlag_deregister_handler, NULL, 0,
```

Comment:
Because of following reasons I am creating a new task.
1) It Involves IPC with Zebra, which can suspend the Thread.
2) In Future if it becomes Heavy processing api, it had performance Impact.

Because of these reasons I Created new Task.

Reply:
1. Then we're just suspending later instead of right now. I don't immediately see what the benefit is.
2. I don't understand this, isn't the register message sent once?

Reply:
Need not be. from PIM-MLAG side, its a Dynamic procedure based on dual active Interface configuration.

Only Idea is caller should not be blocked because of thread suspension. 

I am trying to make it like a asynchronous call, so that caller will never suspended